File: solutions/java/squeaky-clean/1/src/main/java/SqueakyClean.java
class SqueakyClean {
    static String clean(String identifier) {
        StringBuilder chars = new StringBuilder(identifier);

        for ( int i = 0; i < chars.length(); i++ ) {
            char c = chars.charAt(i);
            if ( Character.isWhitespace(c) ) {
                chars.setCharAt(i, '_');
            } else if ( c == '0' ) {
                chars.setCharAt(i, 'o');
            } else if ( c == '3' ) {
                chars.setCharAt(i, 'e');
            } else if ( c == '4' ) {
                chars.setCharAt(i, 'a');
            } else if ( c == '7' ) {
                chars.setCharAt(i, 't');
            } else if ( c == '¡' ) {
                chars.deleteCharAt(i);
                if ( i < chars.length() ) {
                    c = chars.charAt(i);
                    if ( c == '1' ) {
                        chars.setCharAt(i, 'l');
                    }
                }
            } else if ( c == '-' ) {
                chars.deleteCharAt(i);
                if ( i < chars.length() ) {
                    c = chars.charAt(i);
                    chars.setCharAt(i, Character.toUpperCase(c));
                }
            } else if ( ! Character.isLetter(c) ) {
                chars.deleteCharAt(i);
                i--;
            }
        }

        return chars.toString();
    }
}

File: solutions/clojure/squeaky-clean/1/src/squeaky_clean.clj
(ns squeaky-clean
  (:require [clojure.string :as str]))

(defn clean
  "a partial set of utility routines to help a developer clean up
  identifier names."
  [s]
  (def name s)

  ;; Replace any spaces encountered with underscores
  (def name (str/replace name #"\s" "_"))

  ;; Replace control characters with the upper case string "CTRL"
  (def name (str/replace name #"\p{Cc}" "CTRL"))

  ;; Convert kebab-case to camelCase
  (def name (str/replace name #"-(\p{Ll}+)" #(str/capitalize (%1 1))))

  ;; Omit characters that are not letters
  (def name
    (apply str (filter (fn [x] (or (Character/isLetter x) (= \_ x))) name)))

  ;; Omit Greek lower case letters
  (def name (str/replace name #"[α-ω]" ""))

  name
  )
